DietPi-Software | Adding Bitwarden_rs

CactiChameleon9 opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready (I think)
- [ ] Example task 1

**Reference**: https://github.com/MichaIng/DietPi/issues/3724

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- This should all work correctly, and testing seems to work
- Still needs testing on a fresh install if that's required

CactiChameleon9 committed:
Add bitwarden_rs

CactiChameleon9 committed:
Meet merge bot expectations

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5519,6 +5529,36 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+
+                    # Rust from https://rustup.rs/
+                    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --default-toolchain none
+                    source "$HOME/.cargo/env"
```
Don't use `$HOME` here, use `/root` instead to be explicit and failsafe. The script does forcefully run as root user anyway.

CactiChameleon9 replied:
Makes sense

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11615,6 +11655,62 @@ _EOF_
+                    G_EXEC mkdir -p /mnt/dietpi_userdata/bitwarden_rs/data
+
+                    # Make and edit the template config to have no web vault
+                    G_EXEC cp /opt/bitwarden_rs/.env.template /mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env
```
It's better to not overwrite existing environment on reinstall, isn't it? What about:
```
if [[ ! -f '/mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env' ]]
then
	G_EXEC cp /opt/bitwarden_rs/.env.template /mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env
	G_CONFIG_INJECT 'WEB_VAULT_ENABLED=' 'WEB_VAULT_ENABLED=false' /mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env
fi
```

CactiChameleon9 replied:
Yep, that is sensible. I forgot about the reinstall command.

MichaIng replied:
The line above then needs to be removed 😄.

CactiChameleon9 replied:
Oops, fixed

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11615,6 +11655,62 @@ _EOF_
+[Service]
+# The user/group bitwarden_rs is run under. the working directory (see below) should allow write and read access to this user/group
+User=bitwarden_rs
+Group=bitwarden_rs
```
Let's do not define the group here. It is obsolete when `bitwarden_rs` is the primary group of that user, if breaks supplementary group permissions, e.g. adding it to an `ssl` group to grant access to SSL certificates, and if it's not the primary group, then it has been intentionally changed by admin, hence should be respected.

CactiChameleon9 replied:
OK, I half get what that means. The service I just copied from their wiki and removed a few comments not applicable due to us using SQLite and not others

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11615,6 +11655,62 @@ _EOF_
+ProtectSystem=strict
+# Only allow writes to the following directory and set it to the working directory (user and password data are stored here)
+WorkingDirectory=/mnt/dietpi_userdata/bitwarden_rs
+ReadWriteDirectories=/mnt/dietpi_userdata/bitwarden_rs
```
Just to keep an eye on: I'm not sure if `PrivateTmp=true` implies that the service has write permissions to it's own isolated `/tmp` directory (would be reasonable) but it might be possible to add `/tmp` to `ReadWriteDirectories` as well for that.

CactiChameleon9 replied:
Again, I just copied from the bitwarden_rs example service

MichaIng replied:
Ah okay then this should work. A test will show quickly anyway 😉.

CactiChameleon9 committed:
Change $HOME to /root

CactiChameleon9 committed:
Fix overiding config on reinstall like suggested

CactiChameleon9 committed:
Remove the service group requirment

CactiChameleon9 commented:
I have fixed the first 3 issues/suggested changes. The last one ( PrivateTmp=true) I don't know much about, but I think it should be left as it is the [service config they give](https://github.com/dani-garcia/bitwarden_rs/wiki/Setup-as-a-systemd-service) so it should all work with it.

CactiChameleon9 commented:
Also, should I add a config for the web vault and a reverse proxy to be installed by default or should I leave it at minimal? It is quite easy to add the web vault - so it isn't hard to do from our end or the end user's end.

CactiChameleon9 committed:
Fixed a me guessing bash error

CactiChameleon9 committed:
Fix server not starting on startup

CactiChameleon9 commented:
I just made a commit(c8d5f61) to make sure the service starts properly, but do I have to register the service with dietpi-services or something or an I doing it correctly?

CactiChameleon9 committed:
Fix dumb logic mistake with overriding config

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5519,6 +5529,36 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+                    G_EXEC cd /tmp/$G_PROGRAM_NAME
+
+                    # Uninstall rust after compiling
+                    G_EXEC rustup self uninstall -y
```
This will uninstall Rust from `/root/.cargo` right? What do you think about installing it to `/tmp`?
Or keeping Rust/cargo allow easy updates or has at an internal updater or otherwise requires a manual source download anyway (=> `dietpi-software reinstall 183` as easy solution)?

CactiChameleon9 replied:
It uses rust nightly, so it would probably have to re-download anyway (and yes it uninstalls from /root/.cargo). May be able to use /tmp, but I don't see any potential gains since it is removed afterwards anyway.

MichaIng replied:
It would reduce physical disk (SD card) writes and potentially speed up things a little... although CPU/compiling is what takes time here, not the disk reads/writes 😄, so it's only to follow the general practice and reduce SD writes wherever possible to increase lifetime and have temporary files always created at the temporary files directory 🙂.

CactiChameleon9 replied:
Oh yeah, /tmp isn't on the sd card, its a link to the ram.
I will try to see if I can change the install path

MichaIng replied:
There is no explicit option but the users home dir is used by default. A custom path can be force by setting the HOME variable, but the script then prints an ugly error message which as well cannot be muted via quite flag:
```
2020-11-20 18:03:24 root@VM-Buster:~# curl -sSfL https://sh.rustup.rs | HOME=/tmp dash -s -- -q -y --default-toolchain none
info: downloading installer
error: $HOME differs from euid-obtained home directory: you may be using sudo
error: $HOME directory: /tmp
error: euid-obtained home directory: /root
info: profile set to 'default'
info: default host triple is x86_64-unknown-linux-gnu
info: skipping toolchain installation


Rust is installed now. Great!

To get started you need Cargo's bin directory ($HOME/.cargo/bin) in your PATH
environment variable. Next time you log in this will be done
automatically.

To configure your current shell run source $HOME/.cargo/env
```
It could be wrapped into G_EXEC, with eval to allow the pipe:
```
2020-11-20 18:03:16 root@VM-Buster:~# G_EXEC eval 'curl -sSfL https://sh.rustup.rs | HOME=/tmp dash -s -- -y --default-toolchain none'
[  OK  ] eval curl -sSfL https://sh.rustup.rs | HOME=/tmp dash -s -- -y --default-toolchain none
```

CactiChameleon9 replied:
Wow, thanks for doing the research

CactiChameleon9 replied:
I have implemented it and tested installing and uninstalling on my machine. I have not tested the compiling, but it will probably also require the home override like the rustup uninstall

MichaIng replied:
I think `source "/tmp/.cargo/env"` covers the custom Rust dir, but does not hurt.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11615,6 +11655,70 @@ _EOF_
+		    if [[ ! -f '/mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env' ]]
+		    then
+			    G_EXEC cp /opt/bitwarden_rs/.env.template /mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env
+			    G_CONFIG_INJECT 'WEB_VAULT_ENABLED=' 'WEB_VAULT_ENABLED=false' /mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env
```
By default the working dir or users $HOME/data seems to be used as data dir, but would it make sense to set this explicitly to be failsafe?
```
G_CONFIG_INJECT 'DATA_FOLDER=' 'DATA_FOLDER=/mnt/dietpi_userdata/bitwarden_rs' /mnt/dietpi_userdata/bitwarden_rs/bitwarden_rs.env
```
That way the `data` sub dir could be skipped as well.

CactiChameleon9 replied:
I suppose that would make sense - I will add that now.

MichaIng commented:
To make own tests correctly and for our online docs:
- I downloaded the Windows desktop client here: https://bitwarden.com/download/
- Top left settings gear, I added `http://192.168.1.22:8000` (test VMs local IP), then I created a new account.
- Seems to work, and yes it's my own server that the account is created on, as when I stop the service, the client "fails to fetch" when I want to add/change something.
_______________
@CactiChameleon9
> I just made a commit(c8d5f61) to make sure the service starts properly, but do I have to register the service with dietpi-services or something or an I doing it correctly?

Jep, add the service name to this array: https://github.com/CactiChameleon9/DietPi/blob/patch-1/dietpi/dietpi-services#L147

> Also, should I add a config for the web vault and a reverse proxy to be installed by default or should I leave it at minimal? It is quite easy to add the web vault - so it isn't hard to do from our end or the end user's end.

Probably we can add it as optional choice via dietpi.txt setting and/or whiptail question and add instructions to online docs? But actually a web UI is definitely helpful and we install optional ones in case of other software titles as well. Also I see it requires additional download and install steps, so probably we install it by default. It can be disabled easily. I'd skip a reverse proxy, also to match other software installs and as we might add a generic script/instructions to setup reverse proxies web applications that do not run on a regular webserver.

CactiChameleon9 committed:
Add bitarden_rs to the service array

CactiChameleon9 committed:
Add a custom data dir to env file

I also removed the making of the no longer needed data dir

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1071,6 +1071,16 @@ DietPi-Software will decrypt and use it for software installs. You can change it
+                aSOFTWARE_DESC[$software_id]='Unoffical Bitwarden password manager server written in rust'
+                aSOFTWARE_TYPE[$software_id]=0
+                aSOFTWARE_CATEGORY_INDEX[$software_id]=4
+                aSOFTWARE_ONLINEDOC_URL[$software_id]='p=24799#p24799' #idk what to do here
```
We need to create an online documentation page: https://dietpi.com/phpbb/viewtopic.php?p=28177#p28177
And here: https://github.com/MichaIng/DietPi-Docs/blob/dev/docs/software/cloud.md
I'll switch those URLs over to the new docs with an upcoming release.

Client setup and web vault access (when implemented) should be added, or rough instructions (or a link) about how to enable the web vault.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11615,6 +11655,71 @@ _EOF_
+ProtectSystem=strict
+# Only allow writes to the following directory and set it to the working directory (user and password data are stored here)
+WorkingDirectory=/mnt/dietpi_userdata/bitwarden_rs
+ReadWriteDirectories=/mnt/dietpi_userdata/bitwarden_rs
```
This has been btw renamed to "ReadWritePaths" while the old name still works for backwards compatibility: https://github.com/caddyserver/caddy/pull/2798 (Stretch ships systemd v232 so no problems to use the new name in DietPi)
https://www.freedesktop.org/software/systemd/man/systemd.exec.html#ReadWritePaths=

CactiChameleon9 committed:
Change to /tmp for rust install

CactiChameleon9 committed:
Change ReadWriteDirectories to ReadWritePaths

CactiChameleon9 committed:
Add installing the web vault

CactiChameleon9 committed:
add home override for the cargo build

CactiChameleon9 committed:
Double quote to word splitting with wget

CactiChameleon9 commented:
All done with all the changes you suggested, and added the web vault installation

CactiChameleon9 committed:
add a forgotten cd

CactiChameleon9 committed:
add G_EXEC to cd

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11615,6 +11655,76 @@ _EOF_
+                    # Download and extract the web vault
+                    cd /mnt/dietpi_userdata/bitwarden_rs
+                    local version=$(curl -sf "https://api.github.com/repos/dani-garcia/bw_web_builds/releases/latest" | grep -m1 '^[[:blank:]]*"tag_name":' | cut -d \" -f 4)
+                    G_EXEC wget -c "https://github.com/dani-garcia/bw_web_builds/releases/download/$version/bw_web_$version.tar.gz" -O - | tar -xz
```
Since the URL ships a `browser_download_url`, and `Download_Install` automated download and extraction:
```
local fallback_url='https://github.com/dani-garcia/bw_web_builds/releases/download/v2.16.1/bw_web_v2.16.1.tar.gz'
Download_Install "$(curl -sfL 'https://api.github.com/repos/dani-garcia/bw_web_builds/releases/latest' | mawk -F\" '/browser_download_url.*\.tar\.gz\"/{print $4;exit}')" /mnt/dietpi_userdata/bitwarden_rs
```
- The `fallback_url` is used when below curl does not match something, e.g. the file names change or something, then the latest known version is downloaded.

An it does not really make a difference, but maybe this fits better into the install step?

CactiChameleon9 replied:
OK, that makes more sense, the Download_Install does include extraction correct?

MichaIng replied:
Jep, it does it based on the file type for a set of archive types, zip, 7z, tarballs (xz is still missing 😅) + installs DEB packages.

CactiChameleon9 committed:
Change to better method of downloading and extracting

MichaIng committed:
v6.34

+ DietPi-Software | Bitwarden_RS: Align indentation
+ DietPi-Software | Bitwarden_RS: Rename to "Bitwarden_RS" with underscore with matches the official spelling
+ DietPi-Software | Bitwarden_RS: Export HOME override and revert, to simplify it a bid and assure that it is set for all commands and sub commands where required
+ DietPi-Software | Bitwarden_RS: Add fallback URL for main download
+ DietPi-Software | Bitwarden_RS: Fix web vault install where G_CHECK_URL on the final archive fails, and move that to the install step

MichaIng commented:
Hmm, when I want to create an account via web vault, hitting "Send" does not have any effect but produces an error in browser console:
```
core.js:4127 ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'importKey' of null
TypeError: Cannot read property 'importKey' of null
    at e.<anonymous> (webCryptoFunction.service.ts:47)
    at main.cd3fdd91e1a63cf3ead5.js:1
    at Object.next (main.cd3fdd91e1a63cf3ead5.js:1)
    at main.cd3fdd91e1a63cf3ead5.js:1
    at new t (zone.js:913)
    at r (main.cd3fdd91e1a63cf3ead5.js:1)
    at e.pbkdf2 (main.cd3fdd91e1a63cf3ead5.js:1)
    at e.<anonymous> (crypto.service.ts:315)
    at main.cd3fdd91e1a63cf3ead5.js:1
    at Object.next (main.cd3fdd91e1a63cf3ead5.js:1)
    at w (zone.js:832)
    at zone.js:739
    at s (main.cd3fdd91e1a63cf3ead5.js:1)
    at t.invoke (zone.js:386)
    at Object.onInvoke (core.js:28144)
    at t.invoke (zone.js:385)
    at e.run (zone.js:143)
    at zone.js:891
    at t.invokeTask (zone.js:421)
    at Object.onInvokeTask (core.js:28132)
```

When trying to login (just to check whether as well console errors appear), it reveals another, probably related issue, as the web UI itself prints an error:
```
This browser requires HTTPS to use the web vault
Check the bitwarden_rs wiki for details on how to enable it
```

"This" browser is latest Opera dev, so all current Chromium-based browsers might be affected. Chromium Edge shows indeed exactly the same error and behaviour.

Hmm, I'm not keen to install a whole webserver as reverse proxy and creating a trusted standalone certificate is possible but not yet implemented into `dietpi-letsencrypt`, also those would not grant access via plain IP but only domain. Probably we could place a snakeoil certificate, just to make it running, and advise users to create a real one or setup a reverse proxy. I mean a snakeoil certificate is not worse than plain HTTP, as long as you do not feel "safe" just because of the HTTPS protocol.
___________
Initial online docs: https://dietpi.com/phpbb/viewtopic.php?p=28903#p28903

MichaIng committed:
v6.34

+ DietPi-Software | Align methods when downloading from GitHub archive via latest version tag

CactiChameleon9 commented:
That error is very odd - in theory I programmed it to do the exact way I manually installed. I need to test on a blank install. I think a snake-oil cert would be best - however I am not too confident with self signing things.

MichaIng commented:
Known issue: https://github.com/dani-garcia/bitwarden_rs/issues/153
Known and intended by upstream vault: https://github.com/bitwarden/web/issues/254
I'll implement a self-signed certificate generation.

MichaIng committed:
v6.34

+ DietPi-Software | Bitwarden_RS: Web vault requires HTTPS to work with all browsers, so create a self-signed certificate.

MichaIng commented:
Memory + disk (RAMdisk) usage is quite high during compilation:
```
2020-11-25 12:11:28 root@VM-Bullseye:/tmp# du -sm .rustup .cargo
671     .rustup
198     .cargo
2020-11-25 12:11:26 root@VM-Bullseye:/tmp# free -m
              total        used        free      shared  buff/cache   available
Mem:           1989         592         138        1103        1258         153
Swap:             0  
```
Even if we force 2 GiB memory via auto-sized swap file (when it's smaller) and temporarily increase `/tmp` to 2 GiB, it's very close 🤔.
______
Too close, 2 GiB was not sufficient, build of a specific module with high RAM demand failed. I needed to add another GiB to swap space, then it went through.

On a 4 core/thread machine, `cargo` utilises 4 `rustc` processes to compile concurrently, each using ~200 - 500 MiB memory, the last, `bitwarden_rs` itself, requires 700 MiB peak. While this sounds reasonable at first, actually a single `rustc` process can utilise all 4 cores/threads, i.e. 400% CPU usage. So running multiple processes concurrently not only leads to a very high memory usage but in theory does not speed up things but slows down due to parallelism overhead. Probably there is a way to force only a single `rustc` process at a time.
_______
Another issue, with a self-signed certificate, the client fails to access:
```
Failed to load resource: net::ERR_CERT_AUTHORITY_INVALID
```
From the docs: https://bitwarden.com/help/article/install-on-premise/
> - If you are only testing and do not have an SSL certificate, a self-signed certificate can be generated for your installation. Self-signed certificates will not be trusted by Bitwarden client applications so you will need to install this certificate to the trusted store of each device you plan to use Bitwarden with.
>
> - If you choose not to configure Bitwarden with a SSL certificate you must front your installation with a proxy that serves the Bitwarden installation over SSL. HTTPS is required to use Bitwarden. If you try to use Bitwarden without the HTTPS protocol you will get errors.

Hmm here it says, importing the certificate to trusted storage is enough: https://bitwarden.com/help/article/hosting-faqs/#trust-a-private-ca-issued-or-self-signed-certificate-for-bitwarden-client
_______
Okay importing works, at least we're one step further:
```
Failed to load resource: net::ERR_CERT_COMMON_NAME_INVALID
```
So dummy entries are required.
... not even CN, actually SAN is required and needs to match:
```
openssl req -x509 -sha256 -newkey rsa:4096 -days 7200 -reqexts SAN -extensions SAN -nodes -keyout privkey.pem -out cert.pem -subj '/CN=DietPi Bitwarden_RS' -config <(cat /etc/ssl/openssl.cnf <(echo -ne "[SAN]\nsubjectAltName=DNS:$(</etc/hostname),IP:$(mawk 'NR==4' /run/dietpi/.network)"))
```
Only works for local network access via IP or the servers internal hostname, or, if the public FQDN matches the `/etc/hostname` entry, which is generally reasonable.

And of course the certificate needs to be imported into trusted CA space. Probably we need to give an introduction about how to do that on Windows 10 and Linux? Has become even more complicated since it cannot be installed directly from the browser but needs to be copied to file first, then file opened and installed from there...
For removal: `certmgr.msc`

MichaIng committed:
v6.34

+ DietPi-Software | Bitwarden_RS: Add SAN to self-signed TLS certificate, otherwise the official Bitwarden client refuses connection. As well this makes the connection appear "secure" in some browsers, at least after the certificate has been added to the trusted CA storage.

MichaIng committed:
v6.34

+ DietPi-Software | Bidwarden_RS: Assure 2 GiB memory and 1.5 GiB /tmp size for Rust install to /tmp and Bitwarden_RS build
+ DietPi-Software | Bidwarden_RS: Don't run multiple jobs concurrently, as one job can require > 500 MiB memory and can utilize all cores as well
+ DietPi-Software | Bidwarden_RS: Download and execute rustup as separate steps to allow separate error handling, even editing rustup-init.sh, when required, and avoid "eval" where we don't exactly know the scripts content.
+ DietPi-Software | Mycroft AI: Avoid running "dietpi-set_swapfile 1" if less than 100 MiB would be required, as such a small swap file size is now omitted by the script internally, which lead to ~50 MiB swap files on all 2 GiB physical RAM devices in the past.

MichaIng commented:
I'll liberate `dietpi-letsencrypt` in a separate PR to allow retrieving a certificate without having a webserver installed. The code is already there when Minio S3 file server is installed. It makes sense to simply always create a standalone certificate, and, if one or more TLS-capable service is installed, configure those to get a copy of that certificate.

We could automatically apply a Certbot certificate, when found on the system /etc/letsencrypt/live/... but that needs to be thought through as well since this means that the service can only be accessed via the public FQDN (fully qualified domain name), not via local IP or similar, as the certificate is not trusted then. So for local network access, a self-signed certificate can even be seen as a benefit, to allow local network access via local hostnames or IPs while for external access port 443 -> 3000 proxy has the benefit that only ports 80/443 need to be opened to the web. Bitwarden_RS could still run via HTTPS so even local network-internal proxy connection is encrypted.

So generally as non-HTTP with Bitwarden in general doesn't work reliably (strangely works for the Windows desktop client, but not for web vault, and probably not all clients), at least the docs say it causes issues and it totally makes sense to have the connection encrypted that transports the master password for a password management server. But as the ways to set this up all have their pros and cons, trusted CA certificates require a public domain name and ports 80/443 opened, it seems to be best to automatically only setup a simple self-signed certificate based on the best information we can get (local IP and machine hostname) and add further information to our online docs about which steps are required to use that self-signed certificate with a recommendation to setup a real trusted CA signed one ASAP for remote access.

MichaIng commented:
Okay, number of concurrent jobs DOES limit the utilised cores for a single job as well. Probably when allowing 4 concurrent jobs, there are some special jobs (like compiling bitwarden_rs binary itself) that can use and are then allowed to use 4 cores while allowing only 1 concurrent job limits as well the used cores/threads of jobs that could utilise more core 🤔. Final solution: If more than two logic cores are available but less than 3 GiB memory, limited jobs to two, else utilising all available cores is fine.

CactiChameleon9 commented:
> And of course the certificate needs to be imported into trusted CA space. Probably we need to give an introduction about how to do that on Windows 10 and Linux? Has become even more complicated since it cannot be installed directly from the browser but needs to be copied to file first, then file opened and installed from there...
> For removal: `certmgr.msc`

Can a web browser just say accept the risk and continue?

Thanks for doing the cert stuff BTW

Also good research/testing on compiling issues

MichaIng committed:
v6.34

+ DietPi-Software | Bitwarden: Allow using all CPU cores for cargo build when >= 3 GiB memory are available, else limit to 2 cores

MichaIng commented:
> Can a web browser just say accept the risk and continue?

With most browser one can do that, but not with all. But most importantly the Windows desktop Bitwarden client does not accept it without having it imported as trusted CA, and I guess in case of some mobile clients it's similar.
__________
Okay, ready from my side, @CactiChameleon9 are you okay with everything?

CactiChameleon9 commented:
I think its all OK. I haven't actually tested it recently - but I am quite sure it works

MichaIng committed:
v6.34

+ DietPi-Software | Bitwarden_RS: If APT cache and/or archives have been moved to RAM, build can still run into OOM due to the large amount of DEB packages installed as dependencies. Be failsafe and clean both. As the binary cache file is removed, this could slow down further APT installs within the same run, but much less compared to limiting the build to a single thread and Bitwarden_RS is one of the last installs performed due to its location near the end of the install function.

MichaIng commented:
New documentation page is ready as well: https://github.com/MichaIng/DietPi-Docs/pull/253

CactiChameleon9 commented:
Well done

MichaIng commented:
We have already three software titles running at port 8000: WebIOPi, IceCast and Koel, so we should change it to 8001.

MichaIng committed:
v6.34

+ DietPi-Software | Bitwarden_RS: Adjust listening port from 8000 to 8001 to avoid conflict with WebIOPi, IceCast and Koel

MichaIng committed:
v6.34

+ Typo
